### PR TITLE
Implement AST reflection

### PR DESCRIPTION
**Pull request by krame505**
This implements the `reflectAST` operation as described in #212.  Next up is reify.

### CODE REVIEW

**Comment by krame505** [comment]
@tedinski @ericvanwyk Anything that I should fix before merging this?

File: runtime/java/src/common/Node.java [hunk 1]
```diff
@@ -96,6 +96,21 @@ public DecoratedNode decorate() {
...
 	
+	/**
+	 * Used for reflection.
+	 * 
+	 * @return An array of the names of all annotations.
+	 */
+	public abstract String[] getAnnoNames();
+
+	/**
+	 * Used for debugging, stack traces especially.
```
Comment (tedinski) [inline runtime/java/src/common/Node.java]: Text from copy and paste?

**Comment by tedinski** [comment]
General thoughts:

1. Not sure if we should have AST have a `pp`, but for now okay no problem. But I think we could probably avoid the dependency on langutil in the final version?
2. In the end, we probably need tests that do better than compare to string (too brittle?), but obviously I think that's hard to do yet? So ok.

Probably good to merge.

**Comment by ericvanwyk** [comment]
Ok with me.

File: runtime/java/src/common/Node.java (hunk 1 above)
Comment (krame505) [reply runtime/java/src/common/Node.java]: Oops...

**Comment by krame505** [comment]
> 1. Not sure if we should have AST have a pp, but for now okay no problem. But I think we could probably avoid the dependency on langutil in the final version?

Yeah, I suppose those equations could move to aspects in the langutil library.  

> 2. In the end, we probably need tests that do better than compare to string (too brittle?), but obviously I think that's hard to do yet? So ok.

Agreed, we really need a derivable equality typeclass for this...
I'll go ahead and merge.

**Pull request merged**